File: app/lib/services/aws_v4_signer.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

/// Lightweight AWS Signature V4 implementation for S3 requests.
/// Only depends on the `crypto` package (HMAC-SHA256 / SHA-256).
class AwsV4Signer {
  final String accessKeyId;
  final String secretAccessKey;
  final String region;
  final String service;

  AwsV4Signer({
    required this.accessKeyId,
    required this.secretAccessKey,
    required this.region,
    this.service = 's3',
  });

  /// Generate a presigned URL for the given HTTP [method] and [uri].
  /// The URL is valid for [expireSeconds] (default 3600).
  Uri presignUrl({
    required String method,
    required Uri uri,
    int expireSeconds = 3600,
    Map<String, String>? additionalHeaders,
  }) {
    final now = DateTime.now().toUtc();
    final dateStamp = _dateStamp(now);
    final amzDate = _amzDate(now);
    final scope = '$dateStamp/$region/$service/aws4_request';

    final headers = <String, String>{
      'host':
          uri.host +
          (uri.hasPort && uri.port != 443 && uri.port != 80
              ? ':${uri.port}'
              : ''),
      ...?additionalHeaders,
    };
    final signedHeadersList = headers.keys.map((k) => k.toLowerCase()).toList()
      ..sort();
    final signedHeaders = signedHeadersList.join(';');

    final queryParams = Map<String, String>.from(uri.queryParameters);
    queryParams['X-Amz-Algorithm'] = 'AWS4-HMAC-SHA256';
    queryParams['X-Amz-Credential'] = '$accessKeyId/$scope';
    queryParams['X-Amz-Date'] = amzDate;
    queryParams['X-Amz-Expires'] = expireSeconds.toString();
    queryParams['X-Amz-SignedHeaders'] = signedHeaders;

    final sortedParams = Map.fromEntries(
      queryParams.entries.toList()..sort((a, b) => a.key.compareTo(b.key)),
    );
    final canonicalQueryString = sortedParams.entries
        .map((e) => '${_uriEncode(e.key)}=${_uriEncode(e.value)}')
        .join('&');

    final canonicalHeaders = signedHeadersList
        .map(
          (h) =>
              '$h:${headers[h]?.trim() ?? headers[h.toLowerCase()]?.trim() ?? ''}\n',
        )
        .join();

    final canonicalRequest = [
      method.toUpperCase(),
      _uriEncodePath(uri.path.isEmpty ? '/' : uri.path),
      canonicalQueryString,
      canonicalHeaders,
      signedHeaders,
      'UNSIGNED-PAYLOAD',
    ].join('\n');

    final stringToSign = [
      'AWS4-HMAC-SHA256',
      amzDate,
      scope,
      _sha256Hex(utf8.encode(canonicalRequest)),
    ].join('\n');

    final signingKey = _deriveSigningKey(dateStamp);
    final signature = _hmacSha256Hex(signingKey, utf8.encode(stringToSign));

    final finalQuery = '$canonicalQueryString&X-Amz-Signature=$signature';
    return uri.replace(query: finalQuery);
  }

  /// Sign a request by adding Authorization and related headers.
  /// Returns a map of headers to merge into the HTTP request.
  Map<String, String> signRequest({
    required String method,
    required Uri uri,
    Map<String, String>? headers,
    List<int>? body,
  }) {
    final now = DateTime.now().toUtc();
    final dateStamp = _dateStamp(now);
    final amzDate = _amzDate(now);
    final scope = '$dateStamp/$region/$service/aws4_request';

    final payloadHash = body != null && body.isNotEmpty
        ? _sha256Hex(body)
        : _sha256Hex(const <int>[]);

    final allHeaders = <String, String>{
      'host':
          uri.host +
          (uri.hasPort && uri.port != 443 && uri.port != 80
              ? ':${uri.port}'
              : ''),
      'x-amz-date': amzDate,
      'x-amz-content-sha256': payloadHash,
      ...?headers,
    };
    final signedHeadersList =
        allHeaders.keys.map((k) => k.toLowerCase()).toList()..sort();
    final signedHeaders = signedHeadersList.join(';');

    final canonicalQueryString = uri.queryParameters.entries.toList()
      ..sort((a, b) => a.key.compareTo(b.key));
    final queryString = canonicalQueryString
        .map((e) => '${_uriEncode(e.key)}=${_uriEncode(e.value)}')
        .join('&');

    final canonicalHeaders = signedHeadersList
        .map((h) => '$h:${allHeaders[h]?.trim()}\n')
        .join();

    final canonicalRequest = [
      method.toUpperCase(),
      _uriEncodePath(uri.path.isEmpty ? '/' : uri.path),
      queryString,
      canonicalHeaders,
      signedHeaders,
      payloadHash,
    ].join('\n');

    final stringToSign = [
      'AWS4-HMAC-SHA256',
      amzDate,
      scope,
      _sha256Hex(utf8.encode(canonicalRequest)),
    ].join('\n');

    final signingKey = _deriveSigningKey(dateStamp);
    final signature = _hmacSha256Hex(signingKey, utf8.encode(stringToSign));

    return {
      'Authorization':
          'AWS4-HMAC-SHA256 Credential=$accessKeyId/$scope, SignedHeaders=$signedHeaders, Signature=$signature',
      'x-amz-date': amzDate,
      'x-amz-content-sha256': payloadHash,
    };
  }

  // ── Helpers ────────────────────────────────────────────────────────

  Uint8List _deriveSigningKey(String dateStamp) {
    final kDate = _hmacSha256(
      utf8.encode('AWS4$secretAccessKey'),
      utf8.encode(dateStamp),
    );
    final kRegion = _hmacSha256(kDate, utf8.encode(region));
    final kService = _hmacSha256(kRegion, utf8.encode(service));
    return _hmacSha256(kService, utf8.encode('aws4_request'));
  }

  static String _dateStamp(DateTime dt) =>
      '${dt.year}${dt.month.toString().padLeft(2, '0')}${dt.day.toString().padLeft(2, '0')}';

  static String _amzDate(DateTime dt) =>
      '${_dateStamp(dt)}T${dt.hour.toString().padLeft(2, '0')}'
      '${dt.minute.toString().padLeft(2, '0')}${dt.second.toString().padLeft(2, '0')}Z';

  static Uint8List _hmacSha256(List<int> key, List<int> data) {
    final hmac = Hmac(sha256, key);
    return Uint8List.fromList(hmac.convert(data).bytes);
  }

  static String _hmacSha256Hex(List<int> key, List<int> data) {
    final hmac = Hmac(sha256, key);
    return hmac.convert(data).toString();
  }

  static String _sha256Hex(List<int> data) => sha256.convert(data).toString();

  static String _uriEncode(String value) {
    return Uri.encodeComponent(value).replaceAll('+', '%20');
  }

  static String _uriEncodePath(String path) {
    return path.split('/').map((s) => _uriEncode(s)).join('/');
  }
}

File: web/src/lib/services/awsV4Signer.ts
import { hmacSha256 as hmacSha256Async, hmacSha256Hex, sha256Hex } from '@/lib/cryptoHash';

const encoder = new TextEncoder();

async function hmacSha256(key: ArrayBuffer | Uint8Array, data: string): Promise<ArrayBuffer> {
  return hmacSha256Async(key, data);
}

function bufToHex(buf: ArrayBuffer): string {
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

function dateStamp(dt: Date): string {
  return dt.toISOString().slice(0, 10).replace(/-/g, '');
}

function amzDate(dt: Date): string {
  return dt.toISOString().replace(/[-:]/g, '').replace(/\.\d{3}/, '');
}

function uriEncode(value: string): string {
  return encodeURIComponent(value).replace(/%20/g, '%20');
}

function uriEncodePath(path: string): string {
  return path
    .split('/')
    .map((s) => uriEncode(s))
    .join('/');
}

async function deriveSigningKey(
  secretKey: string,
  ds: string,
  region: string,
  service: string,
): Promise<ArrayBuffer> {
  const kDate = await hmacSha256(encoder.encode('AWS4' + secretKey), ds);
  const kRegion = await hmacSha256(kDate, region);
  const kService = await hmacSha256(kRegion, service);
  return hmacSha256(kService, 'aws4_request');
}

export interface PresignOptions {
  method: string;
  url: URL;
  accessKeyId: string;
  secretAccessKey: string;
  region: string;
  service?: string;
  expireSeconds?: number;
}

export async function presignUrl(opts: PresignOptions): Promise<string> {
  const {
    method,
    url,
    accessKeyId,
    secretAccessKey,
    region,
    service = 's3',
    expireSeconds = 3600,
  } = opts;

  const now = new Date();
  const ds = dateStamp(now);
  const amz = amzDate(now);
  const scope = `${ds}/${region}/${service}/aws4_request`;

  const host = url.port && url.port !== '443' && url.port !== '80'
    ? `${url.hostname}:${url.port}`
    : url.hostname;

  const signedHeaders = 'host';

  const params = new URLSearchParams(url.search);
  params.set('X-Amz-Algorithm', 'AWS4-HMAC-SHA256');
  params.set('X-Amz-Credential', `${accessKeyId}/${scope}`);
  params.set('X-Amz-Date', amz);
  params.set('X-Amz-Expires', String(expireSeconds));
  params.set('X-Amz-SignedHeaders', signedHeaders);

  const sortedParams = [...params.entries()].sort((a, b) => a[0].localeCompare(b[0]));
  const canonicalQueryString = sortedParams
    .map(([k, v]) => `${uriEncode(k)}=${uriEncode(v)}`)
    .join('&');

  const canonicalRequest = [
    method.toUpperCase(),
    uriEncodePath(url.pathname || '/'),
    canonicalQueryString,
    `host:${host}\n`,
    signedHeaders,
    'UNSIGNED-PAYLOAD',
  ].join('\n');

  const stringToSign = [
    'AWS4-HMAC-SHA256',
    amz,
    scope,
    await sha256Hex(canonicalRequest),
  ].join('\n');

  const signingKey = await deriveSigningKey(secretAccessKey, ds, region, service);
  const signature = await hmacSha256Hex(signingKey, stringToSign);

  return `${url.origin}${url.pathname}?${canonicalQueryString}&X-Amz-Signature=${signature}`;
}

export interface SignRequestOptions {
  method: string;
  url: URL;
  accessKeyId: string;
  secretAccessKey: string;
  region: string;
  service?: string;
  headers?: Record<string, string>;
  bodyHash?: string;
}

export async function signRequest(opts: SignRequestOptions): Promise<Record<string, string>> {
  const {
    method,
    url,
    accessKeyId,
    secretAccessKey,
    region,
    service = 's3',
    headers: extraHeaders = {},
    bodyHash,
  } = opts;

  const now = new Date();
  const ds = dateStamp(now);
  const amz = amzDate(now);
  const scope = `${ds}/${region}/${service}/aws4_request`;
  const payloadHash = bodyHash ?? await sha256Hex('');

  const host = url.port && url.port !== '443' && url.port !== '80'
    ? `${url.hostname}:${url.port}`
    : url.hostname;

  const allHeaders: Record<string, string> = {
    host,
    'x-amz-date': amz,
    'x-amz-content-sha256': payloadHash,
    ...extraHeaders,
  };

  const signedHeadersList = Object.keys(allHeaders)
    .map((k) => k.toLowerCase())
    .sort();
  const signedHeaders = signedHeadersList.join(';');

  const sortedParams = [...url.searchParams.entries()].sort((a, b) => a[0].localeCompare(b[0]));
  const queryString = sortedParams
    .map(([k, v]) => `${uriEncode(k)}=${uriEncode(v)}`)
    .join('&');

  const canonicalHeaders = signedHeadersList
    .map((h) => `${h}:${(allHeaders[h] ?? '').trim()}\n`)
    .join('');

  const canonicalRequest = [
    method.toUpperCase(),
    uriEncodePath(url.pathname || '/'),
    queryString,
    canonicalHeaders,
    signedHeaders,
    payloadHash,
  ].join('\n');

  const stringToSign = [
    'AWS4-HMAC-SHA256',
    amz,
    scope,
    await sha256Hex(canonicalRequest),
  ].join('\n');

  const signingKey = await deriveSigningKey(secretAccessKey, ds, region, service);
  const signature = await hmacSha256Hex(signingKey, stringToSign);

  return {
    Authorization: `AWS4-HMAC-SHA256 Credential=${accessKeyId}/${scope}, SignedHeaders=${signedHeaders}, Signature=${signature}`,
    'x-amz-date': amz,
    'x-amz-content-sha256': payloadHash,
  };
}
